lib/activerecord/delay_touching/state.rb

```diff
@@ -49,6 +49,16 @@ def clear_records
         @already_updated_records.clear
       end
 
+      # If we don't do this then an infinite loop is possible due to how 
+      # Set#subtract and ActiveRecord::Core#== work with the in memory changes 
+      # from a rollback and active record sync
+      def remove_unpersisted_records!
+        @records.each do |attr, set|
+          set.instance_variable_get(:@hash).rehash
```

Comment:
(Same as what I commented on your local PR.)

I think for rehashing on L57 to have an effect, it relies on ActiveRecord's sync_with_transaction_state being invoked upstream. From the AR source, that method is lazily executed, and if it's not run, I think rehash will have no effect. (See below.)

If that is the case, I think it's clearer to invoke explicitly here (as suggested in #14), whether it's via the private method itself or some public API.

For example:
```ruby
# SETUP

records = Set.new
ActiveRecord::Base.transaction do
  records << Micropost.create!(post_type: 'topic', post_category: 'log', content: 'test')
  raise ActiveRecord::Rollback
end

# BAD

records.instance_variable_get(:@hash).rehash # micropost still has ID set (no effect)
records.keep_if(&:persisted?) # this keeps the micropost above (since key is out of sync)

# GOOD

records.each {|record| record.send(:sync_with_transaction_state) }
records.instance_variable_get(:@hash).rehash # now rehashing has an effect
records.keep_if(&:persisted?)  # works as expected
```

If this does work, I think it's only incidental, and that something more explicit would be better.

Reply:
Here's the example I'm working with:

```
ActiveRecord::Base.delay_touching do
  ActiveRecord::Base.transaction do
    # write and touch any new record
    a = Order.create!
    a.touch

    b = Order.create!

    User.last.touch
    Order.first.touch
    raise ActiveRecord::Rollback
  end

  # results in an infinite loop
end
```

The result should be:

  - [x] No infinite loop
  - [x] A single touch update to the last user
  - [x] A single touch update to the first order

So far so good. Here are the logs:


```
(0.3ms)  BEGIN

SQL (0.5ms)  INSERT INTO "orders" ("status", "number", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["status", "pending"], ["number", 27525204], ["created_at", "2018-08-13 20:06:23.381004"], ["updated_at", "2018-08-13 20:06:23.381004"]]

Order Exists (0.5ms)  SELECT  1 AS one FROM "orders" WHERE "orders"."number" = 36529264 LIMIT 1

SQL (0.8ms)  INSERT INTO "orders" ("status", "number", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["status", "pending"], ["number", 36529264], ["created_at", "2018-08-13 20:06:23.404695"], ["updated_at", "2018-08-13 20:06:23.404695"]]

User Load (1.2ms)  SELECT  "users".* FROM "users"  ORDER BY "users"."id" DESC LIMIT 1

Order Load (1.0ms)  SELECT  "orders".* FROM "orders"  ORDER BY "orders"."id" DESC LIMIT 1

(0.3ms)  ROLLBACK

remove_unpersisted_records

ATTR

SET #<Set: {#<Order id: nil, number: 27525204>, #<User id: 5335560>, #<Order id: 31426539, number: 36529264>}>

REHASH {#<Order id: nil, number: 27525204>=>true, #<User id: 5335560>=>true, #<Order id: 31426539, number: 36529264>=>true}

STRIP #<Set: {#<User id: 5335560>, #<Order id: 31426539, number: 36529264>}>

EMPTY? false

(0.3ms)  BEGIN

SQL (0.9ms)  UPDATE "users" SET "updated_at" = '2018-08-13 20:06:23.440097' WHERE "users"."id" = 5335560

SQL (0.6ms)  UPDATE "orders" SET "updated_at" = '2018-08-13 20:06:23.443024' WHERE "orders"."id" = 31426539

(0.6ms)  COMMIT
```